Storage/src/Connection/Rest.php

```diff
@@ -285,6 +285,24 @@ public function insertObject(array $args = [])
     {
         $args = $this->resolveUploadOptions($args);
 
+        // Passing on precondition if present
+        if (isset($args['ifGenerationMatch'])) {
```

Comment:
So, I tested uploading a file multiple times with `$options['ifGenerationMatch'] = 0;` and it never threw an error.

When we pass `ifGenerationMatch = 0`, we expect a failure(This also needs to be in system tests). So, probably these options are not being passed properly.

According to https://cloud.google.com/storage/docs/json_api/v1/objects/insert these preconditions need to be passed in the query params for uploads.

Can you check if that's happening. According to my tests it isn't.

When I used this small snippet inside `insertObject` below the declaration of `$uriParams`, everything works.

```
if(isset($args['ifGenerationMatch'])) {
     $uriParams['query']['ifGenerationMatch'] = $args['ifGenerationMatch'];
}
```

However, is this the best place to do this or the Uploader classes, this is something we need to decide.

Reply:
So yes, missed this part in Unit testing thus we got error while actual prod testing.
Fixed it in [7b832e7](https://github.com/googleapis/google-cloud-php/pull/5730/commits/7b832e7234ed254a8f2d206bc12446498d726c61)

As for `is this the best place to do`, yes it seems this is the best place as this is the existing way to pass query params in insertObject method. This is different from others because of the different way of `Request` object getting generated here as compared to other API send requests in `Rest.php` file.